### PR TITLE
Recording extension: fixed stop recording bugs on simulator

### PR DESCRIPTION
**Pull request by srietkerk**
This PR has everything stop recording!! 

- Fixed the thing that Joey mentioned that if there was a failure anywhere in getting the user media or anything else in `record`, we want to make sure the `recordingState` reflects as such. If there are errors or the user denies access to the microphone, `currentlyRecording = false` is now set. (https://github.com/microsoft/pxt-microbit/pull/5082#discussion_r1178573758)

- Audio and recording now stops when the simulator stop event is fired; closes https://github.com/microsoft/pxt-microbit/issues/5089

- If `play` is triggered before the recording takes the length of the timeout, the timeout is cleared and recorder is stopped. I've added 100ms of buffer time for the recording to populate after recorder is stopped, which then allows the newly recorded audio clip to play immediately. Closes https://github.com/microsoft/pxt-microbit/issues/5092

- When `record` is done, the indicator on the browser tab now reflects that recording has stopped. Closes https://github.com/microsoft/pxt-microbit/issues/5091

- The simulator fully mimics what the hardware does with `record` and `play`. This is non-blocking logic. 

You can play with all these things here: https://makecode.microbit.org/app/c2c6876a5e3af38fcc17f5cb4f21e47c388621db-27857f50af

### CODE REVIEW

**Comment by abchatra** [comment]
🏆

File: sim/state/record-audio.ts [hunk 1]
```diff
@@ -7,65 +7,133 @@ namespace pxsim  {
...
+                    this.audioPlaying = false;
+                });
+            }
+        }
     }
 }
 namespace pxsim.record {
+
+    function stopRecording(b: DalBoard): void {
+        b.recordingState.recorder.stop();
```
Comment (kimprice) [inline sim/state/record-audio.ts:28]: Is it safe to assume `recordingState` has been initialized?

File: sim/state/record-audio.ts [hunk 2]
```diff
@@ -7,65 +7,133 @@ namespace pxsim  {
...
+
+        initListeners = () => {
+            if (this.recording) {
+                this.recording.addEventListener("play", () => {
+                    this.audioPlaying = true;
+                });
+
+                this.recording.addEventListener("ended", () => {
+                    this.audioPlaying = false;
+                });
```
Comment (kimprice) [inline sim/state/record-audio.ts:20]: Should we have clean up code to remove the event listeners?

File: sim/state/record-audio.ts (hunk 1 above)
Comment (srietkerk) [reply sim/state/record-audio.ts:28]: I'm pretty sure. If the board is initialized, this line will get called https://github.com/microsoft/pxt-microbit/blob/ed40af178811c7e6a40febbbf0dacd91a3db53bd/sim/dalboard.ts#L102 and thus initialize the state. So as long as the board exists, which is assumed in the contexts in which we call `stopRecording`, `recordingState` should be initialized.

File: sim/state/record-audio.ts (hunk 2 above)
Comment (kimprice) [reply sim/state/record-audio.ts:20]: Yea, although I'm not sure where you should call it.

File: sim/state/record-audio.ts [hunk 3]
```diff
@@ -7,65 +7,139 @@ namespace pxsim  {
...
+                    runtime.queueDisplayUpdate();
+                }, 4000)
+
+                b.recordingState.recorder.ondataavailable = (e: BlobEvent) => {
+                    b.recordingState.chunks.push(e.data);
                 }
 
-            } else {
-                console.log("getUserMedia not supported on your browser!");
+                b.recordingState.recorder.onstop = () => {
```
Comment (kimprice) [inline sim/state/record-audio.ts:77]: nit: how come `onstop` isn't camel cased?

Comment (kimprice) [reply sim/state/record-audio.ts:77]: oh is this just MediaRecorder?

**Review by kimprice — APPROVED** [review]
LGTM!

File: sim/state/record-audio.ts (hunk 3 above)
Comment (srietkerk) [reply sim/state/record-audio.ts:77]: Yeah, its lowercase for event handlers. If I try to use camel case, typescript gives me an error. It looks like this is just a pattern for a built-in objects that have event handlers/inherit from `EventTarget` in JavaScript. A little blurb from the Mozilla docs that I got help from:
image

**Pull request merged**